drivers: flash: Add Infineon CAT1 Flash driver

sreeramIfx opened the pull request:
- Added initial version of Infineon CAT1 Flash driver
- Added binding file for infineon,cat1-flash-controller.yaml
- Added overlays for subsys/nvs and drivers/flash_shell to support cy8cproto_063_ble, cy8cproto_062_4343w boards
- Defined erase-block-size in PSoC6 MPN dtsi.

de-nordic commented on drivers/flash/flash_ifx_cat1.c:
```diff
@@ -0,0 +1,193 @@
+	uint8_t local_buffer[PAGE_LEN];
+
+	/* Start address of page */
+	uint32_t page_start_addr = (write_addr >> PAGE_LEN_P) << PAGE_LEN_P;
```
I do not think that write address should be corrected when it does not align with page size. User may wrongly set start address but that should be rater reported as error, otherwise this just removes some random stuff - something we may suspect user for, but driver should not do.
And the check is in wrong place, because, I assume, that incorrect write address may be already passed to `ifx_cat1_flash_write`

sreeramIfx replied:
Per the below documentation for the flash_write, its not necessary that the source address is aligned to the page size. Also, we read the entire page, modify the bytes to be updated and then write the entire page back, so we are not affecting the existing data..

de-nordic replied:
> Per the below documentation for the flash_write, its not necessary that the source address is aligned to the page size. Also, we read the entire page, modify the bytes to be updated and then write the entire page back, so we are not affecting the existing data..

Yes that is correct, this means that offset of source may be anything and driver should be able to deal with this. For example if you have some DMA support that only reads from uint32 aligned addresses and user gives you `offset % 4 != 0` it is drivers job to copy data to intermediate buffer and deal with this.
But when user requests write flash write offset to be something driver may not deal with, driver should report error, not correct that.

de-nordic replied:
This is from your code
```
static struct flash_parameters ifx_cat1_flash_parameters = {
	.write_block_size = DT_PROP(SOC_NV_FLASH_NODE, write_block_size),
	.erase_value = 0x00,
};
```
so as far as I can tell your device requires write alignment defined by `write-block-size property`, not `PAGE_LEN`, which seems to be `erase-page-size` property.
The `write-block-size` property, as defined here dts/arm/infineon/psoc6/mpns/CY8C624ABZI_S2D44.dtsi, is 4 for the device. Which means that user, while calling the `flash_write` should provide 4 bytes of data with 4 byte alignment in target offset, otherwise error. I assume, by lack of length parameter for `cyhal_flash_write` that there is error in dtsi and write alignment should be equal to erase page size (512).
Your driver call behaves differently than expected from the documentation, because it corrects destination offset, something it should not do.
Does your device allow rewrites of already written bytes?
You provide erase function so I guess that erase is needed between writes?

sreeramIfx replied:
I misunderstood the documentation for the flash_write API, I have now updated the write-block-size to 512 and removed the logic to correct the offset.

When testing this using the nvs application, I have realized that there is the following check in the nvs sub system that restricts write-block-size(NVS_BLOCK_SIZE) to 32. Would you know the reason for this?

```
    /* check that the write block size is supported */
    if (write_block_size > NVS_BLOCK_SIZE || write_block_size == 0) {
        LOG_ERR("Unsupported write block size");
        return -EINVAL;
    } 
```

de-nordic replied:
Because NVS record, or whatever you call it, is of certain size. If your write alignment is higher than that block, this means you would waste flash using NVS, for example if NVS block is 32 bytes and you write block is 512 bytes, you would have to use 512 size block entirely to store the 32 bytes.

sreeramIfx replied:
so should we increase the NVS_BLOCK_SIZE to 512 when using our board?

sreeramIfx committed:
soc: infineon_cat1: Support for .cy_ramfunc

- Add cy_ramfunc in to RAMFUNC_SECTION section

Signed-off-by: Sreeram Tatapudi <sreeram.praveen@infineon.com>

de-nordic commented on drivers/flash/flash_ifx_cat1.c:
```diff
@@ -0,0 +1,229 @@
+#define DT_DRV_COMPAT	  infineon_cat1_flash_controller
+#define SOC_NV_FLASH_NODE DT_PARENT(DT_INST(0, fixed_partitions))
+
+#define PAGE_LEN   DT_PROP(SOC_NV_FLASH_NODE, erase_block_size)
+#define PAGE_LEN_P 9
```
This is not ok. `PAGE_LEN` is picked from DTSI, but `PAGE_LEN_P` is hardcoded and independent from `PAGE_LEN`.
I have seen that the `PAGE_LEN_P` is used for masking offsets; if you want to have a mask rather do something like
`PAGE_MASK ((uint32_t)-1) & ~(PAGE_LEN - 1))`.

de-nordic commented on drivers/flash/flash_ifx_cat1.c:
```diff
@@ -0,0 +1,229 @@
+
+static struct flash_parameters ifx_cat1_flash_parameters = {
+	.write_block_size = DT_PROP(SOC_NV_FLASH_NODE, write_block_size),
+	.erase_value = 0x00,
```
It this correct? Shouldn't `cyhal_flash_get_info` be used to obtain the value?

sreeramIfx replied:
we have the same values in both device tree and cyhal_flash_get_info.. so both will give the same values..

sreeramIfx committed:
drivers: flash: Add Infineon CAT1 Flash driver

- Added initial version of Infineon CAT1 Flash driver
- Added binding file for infineon,cat1-flash-controller.yaml
- Added overlays for subsys/nvs and drivers/flash_shell
to support cy8cproto_063_ble, cy8cproto_062_4343w boards
- Defined erase-block-size in PSoC6 MPN dtsi.

Signed-off-by: Sreeram Tatapudi <sreeram.praveen@infineon.com>

de-nordic commented on drivers/flash/flash_ifx_cat1.c:
```diff
@@ -0,0 +1,190 @@
+
+	cy_rslt_t rslt = cyhal_flash_read(&data->flash_obj, address + config->base_addr,
+					  (uint8_t *)buffer, length);
+	if (rslt != CY_RSLT_SUCCESS) {
+		flash_ifx_sem_give(dev);
+		return -EIO;
+	}
```
Worth at least `LOG_DBG` or `LOG_ERR` the `rslt` here, unless the `cyhal_flash_read` does some logging.
If you have checked everything to this point and the only reason for this function to fail is actual problem somewhere under `cyhal_flash_read`, returning just -EIO gives no info for developer.

de-nordic commented on drivers/flash/flash_ifx_cat1.c:
```diff
@@ -0,0 +1,190 @@
+	uint32_t erase_offset = config->base_addr + (uint32_t)offset;
+	cy_rslt_t rslt;
+
+	if ((offset > 0) && ((offset % PAGE_LEN) != 0)) {
```
`if (offset < 0 || (offset % PAGE_LEN) !=0)` ?

de-nordic commented on drivers/flash/flash_ifx_cat1.c:
```diff
@@ -0,0 +1,190 @@
+		return -EINVAL;
+	}
+
+	if (((erase_offset + size) < config->max_addr) && ((size % PAGE_LEN) != 0)) {
+		return -EINVAL;
+	}
```
`(erase_offset + size) < config->max_addr)` isn't this actually good?
Shouldn't `&&` be `||`?

de-nordic commented on drivers/flash/flash_ifx_cat1.c:
```diff
@@ -0,0 +1,190 @@
+	struct ifx_cat1_flash_data *data = dev->data;
+	cy_rslt_t rslt = cyhal_flash_init(&data->flash_obj);
+
+	if (rslt != CY_RSLT_SUCCESS) {
+		LOG_ERR("Failed to init FLASH hal driver");
```
`rslt` in the log message.

de-nordic commented on drivers/flash/flash_ifx_cat1.c:
```diff
@@ -0,0 +1,190 @@
+	struct ifx_cat1_flash_data *dev_data = dev->data;
+	const struct ifx_cat1_flash_config *dev_config = dev->config;
+	uint32_t write_offset = dev_config->base_addr + (uint32_t)offset;
+	uint8_t *ptr = (uint8_t *)data;
```
That is begging for warning on dropping `const` qualifier.

de-nordic commented on drivers/flash/flash_ifx_cat1.c:
```diff
@@ -0,0 +1,204 @@
+		return -EINVAL;
+	}
+
+	if (data_len == 0) {
+		return 0;
+	}
```
Note for future: do short tests first. 
Here if `data_len` is 0, there are 3 other arg tests above it before we even get to this return.
Other thing is that this is 'no-work' test, which means that even if we have bad args to other function parameters, as tested above, there should be no error, because we are not going to do anything anyway, which means that bad args would not harm us anyway.

sreeramIfx replied:
updated to check the data_len first..

de-nordic reviewed (dismissed):
Looks OK. Thanks.
